File: src/main/java/com/bazbatlabs/smashballs/models/Color.java
package com.bazbatlabs.smashballs.models;

public final class Color {

    public static final Color RED = new Color(1.0f, 0.0f, 0.0f, 1.0f);
    public static final Color GREEN = new Color(0.0f, 1.0f, 0.0f, 1.0f);
    public static final Color BLUE = new Color(0.0f, 0.0f, 1.0f, 1.0f);
    public static final Color YELLOW = new Color(1.0f, 1.0f, 0.0f, 1.0f);
    public static final Color WHITE = new Color(1.0f, 1.0f, 1.0f, 1.0f);
    public static final Color LIGHT_GRAY = new Color(0.66f, 0.66f, 0.66f, 1.0f);
    public static final Color DARK_GRAY = new Color(0.33f, 0.33f, 0.33f, 1.0f);
    public static final Color BLACK = new Color(0.0f, 0.0f, 0.0f, 1.0f);
    public static final Color CLEAR = new Color(0.0f, 0.0f, 0.0f, 0.0f);

    public final float r;
    public final float g;
    public final float b;
    public final float a;

    public Color(float r, float g, float b, float a) {
        this.r = r;
        this.g = g;
        this.b = b;
        this.a = a;
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Color)) {
            return false;
        }
        Color that = (Color)o;

        return (this.r == that.r) && (this.g == that.g) &&
            (this.b == that.b) && (this.a == that.a);
    }

    @Override
    public int hashCode() {
        int result = 17;
        result = 31 * result + Float.floatToIntBits(r);
        result = 31 * result + Float.floatToIntBits(g);
        result = 31 * result + Float.floatToIntBits(b);
        result = 31 * result + Float.floatToIntBits(a);
        return result;
    }

    @Override
    public String toString() {
        return String.format("(r: %.3f, g: %.3f, b: %.3f, a: %.3f)", r, g, b, a);
    }
}

File: src/main/scala/models/Color.scala
package com.bazbatlabs.bricksmash.models

class Color(val r: Float, val g: Float, val b: Float, val a: Float) {

  override def equals(other: Any): Boolean =
    other match {
      case that: Color =>
        (that canEqual this) &&
        r == that.r && g == that.g &&
        b == that.b && a == that.a
      case _ => false
    }

  def canEqual(other: Any): Boolean = other.isInstanceOf[Color]

  override def hashCode: Int = {
    var result = 17
    result = 31 * result + java.lang.Float.floatToIntBits(r)
    result = 31 * result + java.lang.Float.floatToIntBits(g)
    result = 31 * result + java.lang.Float.floatToIntBits(b)
    31 * result + java.lang.Float.floatToIntBits(a)
  }

  override def toString = f"r: $r%.3f, g: $g%.3f, b: $b%.3f, a: $a%.3f"
}

object Color {
  def apply(r: Float, g: Float, b: Float, a: Float) = new Color(r, g, b, a)

  val RED = Color(1.0f, 0.0f, 0.0f, 1.0f)
  val GREEN = Color(0.0f, 1.0f, 0.0f, 1.0f)
  val BLUE = Color(0.0f, 0.0f, 1.0f, 1.0f)
  val YELLOW = Color(1.0f, 1.0f, 0.0f, 1.0f)
  val WHITE = Color(1.0f, 1.0f, 1.0f, 1.0f)
  val LIGHT_GRAY = Color(0.66f, 0.66f, 0.66f, 1.0f)
  val DARK_GRAY = Color(0.33f, 0.33f, 0.33f, 1.0f)
  val BLACK = Color(0.0f, 0.0f, 0.0f, 1.0f)
  val CLEAR = Color(0.0f, 0.0f, 0.0f, 0.0f)
}
